designs/spot-consolidation.md

```diff
@@ -0,0 +1,112 @@
+
+All consolidation actions contain an inherent tradeoff between price and availability. Ideally, we would like to define an algorithm that enables us to calculate some value that represents an instance type that fits our pods with price and availability optimized. Unfortunately, cloud providers don't surface capacity pool information (availability) for spot, meaning that we need to define availability by some heuristic to make this optimization work. This document explores approaches to enable consolidation for spot instances without regressing to lowest price.
+
+## Design Options
+
+### 1: Minimum Flexibility [Recommended] 
+
+Karpenter will replace a spot node with a cheaper spot node for both single and multi-node consolidation. Cheaper spot instance types are selected with the `price-capacity-optimized` strategy and often the cheapest spot instance type is not launched due to the likelihood of interruption. Karpenter would replace spot node for single-node consolidation only if there are more than 15 potential replacement spot instance types. Also, Karpenter would have to limit its flexibility to this threshold while launching replacements with consolidation to avoid repeat interruptions, e.g. "walking down the ladder". Since on-demand capacity uses the lowest price strategy, its behavior is unaffected if these rules are applied equally to both spot and on-demand launches.
+
+Minimum flexibility of 15 spot instance types is decided after an analysis done on the flexiblity of AWS customers request in the launch path. Karpenter does not have the same instance type flexibility requirement for multi-node spot-to-spot consolidations (many nodes to 1 node) since doing so without requiring flexibility won't lead to "race to the bottom" scenarios. It is also to prevent inconsistent behavior between consolidation of mixed capacity type [od,spot] to [spot] and spot capacity type [spot,spot] to [spot] since we already support mixed capacity type multi-node consolidation to spot today without any additional conditions. And finally, the limits is placed only for launching replacements with consolidation instead of the entire launch path inorder to unblock scenarios where a user doesn't currently have enough flexibility with spot. As a result, they will get their intial launch but then they just won't continually consolidate after that launch and the trade-off would be that the user with enough flexibility in their initial requests might get an immediate consolidation after the intial launch due to the way that we are constraining ourselves during the consolidation.
```

Comment:
```suggestion
Minimum flexibility of 15 spot instance types is decided after an analysis done on the flexiblity of AWS customers request in the launch path. Karpenter does not have the same instance type flexibility requirement for multi-node spot-to-spot consolidations (many nodes to 1 node) since doing so without requiring flexibility won't lead to "race to the bottom" scenarios. It is also to prevent inconsistent behavior between consolidation of mixed capacity type [od,spot] to [spot] and spot capacity type [spot,spot] to [spot] since we already support mixed capacity type multi-node consolidation to spot today without any additional conditions.

Finally, a limit for the number of instance types in the launch path is placed only on consolidation instead of the entire launch path in the provisioning loop in order to unblock scenarios where a user doesn't have enough flexibility with spot specified in their NodePool. As a result, their initial launch would succeed but would get no consolidations after their initial launch.
```

nit: Just trying to clarify the point that you are trying to callout here

Reply:
Updated.